Location export

mkangia opened the pull request:
Also merge submodule https://github.com/dimagi/commcarehq-prelogin/pull/106 along

mkangia committed:
attempt 1

mkangia committed:
Add new classes for filter and filter view instead of trying to configure/break earlier ones [skip ci]

mkangia committed:
cleanup [skip ci]

mkangia committed:
attempt contd

mkangia committed:
Update filters for new system

dimagimon commented on corehq/apps/export/views.py:
```diff
@@ -11,6 +11,8 @@
 
 from corehq.apps.export.export import get_export_download, get_export_size
+from corehq.apps.locations.permissions import location_safe
+from corehq.apps.reports.filters.users import LocationRestrictedMobileWorkerFilter, ExpandedMobileWorkerFilter
```
F401 'ExpandedMobileWorkerFilter' imported but unused

dimagimon commented on corehq/apps/export/views.py:
```diff
@@ -1703,16 +1710,21 @@ def _check_export_size(self, export_instances, filters):
+    @property
+    def page_context(self):
+        parent_context = super(DownloadNewFormExportView, self).page_context
+        parent_context['new_export_filters'] = LocationRestrictedMobileWorkerFilter(self.request, self.request.domain).render()
```
E501 line too long (127 > 115 characters)

dimagimon commented on corehq/apps/hqwebapp/templatetags/hq_shared_tags.py:
```diff
@@ -531,3 +531,9 @@ def url_replace(context, field, value):
+
+@register.filter
+def view_pdb(element):
+    import pdb; pdb.set_trace()
```
E702 multiple statements on one line (semicolon)

benrudolph replied:
you should use `ipdb` 😺  so much nicer: https://pypi.python.org/pypi/ipdb

mkangia replied:
Thanks @benrudolph !!

dimagimon commented on corehq/apps/reports/filters/urls.py:
```diff
@@ -1,11 +1,12 @@
 
 urlpatterns = patterns('',
                        url(r'^emwf_options/$', EmwfOptionsView.as_view(), name='emwf_options'),
+                       url(r'^new_emwf_options/$', LocationRestrictedEmwfOptionsView.as_view(), name='new_emwf_options'),
```
E501 line too long (121 > 115 characters)

esoergel commented on corehq/apps/export/views.py:
```diff
@@ -710,8 +713,12 @@ def _process_filters_and_specs(self, in_data):
         filter_form_data, export_specs = self._get_form_data_and_specs(in_data)
+        import re
+        regex = re.compile('(emw=){1}([^&]*)(&){0,1}')
+        matches = regex.findall(filter_form_data['emw'])
```
ha!  nice

mkangia replied:
Ugly cheat though. Not so happy with it.

esoergel replied:
I'm not sure exactly what comes through as `filter_form_data`, but my guess is a string that looks like a URL querystring?  If that's the case, you could look in to how django parses those to generate the `request.GET` dictionary thing.  I bet there's some function from django you could import to break these up more intuitively.

esoergel commented on corehq/apps/locations/models.py:
```diff
@@ -274,6 +274,13 @@ def couch_locations(self, wrapped=True):
+            return self
+
+        users_location = user.get_sql_location(domain)
+        return self & users_location.get_descendants(include_self=True)
```
I'd like to see some tests for this, since it's a pretty complex chain of filters.

esoergel reviewed (commented):
Looks good overall, I'm pleased to see how small a footprint this has.

mkangia commented on corehq/apps/export/filters.py:
```diff
@@ -192,7 +193,7 @@ def to_es_filter(self):
 
     def __init__(self, user_types):
-        _assert_user_types(user_types)
+        # _assert_user_types(user_types)
```
Nope. But need to amend this as some check fails in there.

mkangia commented on corehq/apps/export/forms.py:
```diff
@@ -405,11 +406,12 @@ def get_edit_url(self, export):
-            self._get_user_filter()
+            self._get_group_filter(mobile_user_and_group_slugs),
+            self._get_user_filter(mobile_user_and_group_slugs),
+            self._get_user_ids_filter(mobile_user_and_group_slugs)
```
Location are just shown in the views as of now. Filtering on their basis is yet to be implemented.

Just making less changes and keeping things close to how they were. Will look for improvements and breaking the architecture next.

esoergel replied:
Makes sense.  I like a good prefactor with no functional changes before the feature work.

mkangia commented on corehq/apps/export/forms.py:
```diff
@@ -418,22 +420,46 @@ def _get_datespan_filter(self):
 
-    def _get_user_filter(self):
-        group = self.cleaned_data['group']
+    def _get_user_filter(self, mobile_user_and_group_slugs):
```
👍

mkangia committed:
add filtering by location

mkangia committed:
Fix issue when parsing user type and filter by fetching user_ids for locations

mkangia committed:
Minor Improvements

mkangia committed:
Add filters for Cases

dimagimon commented on corehq/apps/export/filters.py:
```diff
@@ -201,3 +203,11 @@ def to_es_filter(self):
+        self.location_ids = location_ids
+
+    def to_es_filter(self):
+        return location(self.location_ids)
```
W292 no newline at end of file

mkangia committed:
Remove unused code and unwanted changes, group->group_ids,

esoergel commented:
@mkangia FYI I just merged https://github.com/dimagi/commcare-hq/pull/13476, which also needed `accessible_to_user` available as a queryset method.  Farid pulled it from your PR and added some tests, so you may want to merge in or rebase master.

mkangia commented:
Thanks @esoergel for co-ordinating this. Will take care of it.
Thanks @proteusvacuum for adding tests 👍

mkangia committed:
revisit case export changes to use CaseListFilter [skip ci]

mkangia committed:
Filter forms for users irrespective if group is selected or not since now specific users can be selected along with groups as well

mkangia committed:
Code Improvment: Move responsibilities to classes

mkangia committed:
Remove commented code [skip ci]

mkangia committed:
Move changes to new filter and rename it to EmwfFilterFormExport [skip ci]

mkangia committed:
Override dispatch for views to make them location_safe and override _process_filters_and_specs

esoergel commented:
@mkangia @sheelio regarding the expanded mobile worker filter, what should restricted users see?  Is there a written spec for that?  There are user types, groups, users, and locations, and I imagine they would all be restricted in some way.  Here's my guess at the expected behavior:

A normal user would see everything:
```
[demo_user]
[admin]
[Unknown Users]
USA [location]
India [location]
Canada [location]
My Group [group]
ethan [user]
manish [user]
sheel [user]
```

A restricted user assigned to USA would see:
```
USA [location]
ethan [user]
```
That is, locations at our below their location, users at those locations, and no groups or user types.  Does that sound right?

esoergel commented on corehq/apps/export/forms.py:
```diff
@@ -501,23 +583,43 @@ def _get_datespan_filter(self):
+
+        location_ids = ExpandedMobileWorkerFilter.selected_location_ids(mobile_user_and_group_slugs)
+        if location_ids:
+            default_filters.append(self._get_location_filter(location_ids))
```
Cases can also be owned by locations themselves, so I think this needs a `OwnerTypeFilter(location_ids)` somewhere too.

esoergel commented on corehq/apps/export/forms.py:
```diff
@@ -501,23 +583,43 @@ def _get_datespan_filter(self):
+            OwnerFilter(owner_filter_ids),
+            LastModifiedByFilter(last_modified_filter_ids),
+            *self._get_group_independent_filters(mobile_user_and_group_slugs)
+        )]
```
A lot of this is similar to [`CaseListMixin.case_owners`](https://github.com/dimagi/commcare-hq/blob/c0dc05c0e5859dc7cd1152fc2927512c4c9fae49/corehq/apps/reports/standard/cases/basic.py#L114), which also uses the EMWF to filter cases.  I'm not sure if reusing code is doable (maybe!), but it could be useful at least as a reference.

esoergel reviewed (commented):
What's the long-term plan with export filters?  Are you going to try to make all form and case exports use the EMWF?  Is that out of scope for this PR?  There's a bit of complexity here around supporting both versions, so it's making me wonder how difficult it'd be to drop the old filters entirely.

sheelio commented:
@mkangia @esoergel 

If the HQ user is restricted by location, the following changes should be made to the filter:
- The default options on the filter the assigned locations of the user
- The list of available groups is accessible groups
- The list of available locations is accessible locations
- The list of available mobile workers is accessible mobile workers
- The list of available web users is accessible web users

(Tried to make it clearer in the spec doc).

mkangia committed:
Avoid removing fields for form as layout is re-initialised

mkangia committed:
skip layout assignment in base class when not needed [skip ci]

mkangia commented on corehq/apps/export/forms.py:
```diff
@@ -436,6 +439,80 @@ def get_multimedia_task_kwargs(self, export, download_id):
+
+    def _get_group(self, mobile_user_and_group_slugs):
+        # Override to fetch params from url and not form_data
+        group_ids = CaseListFilter.selected_group_ids(mobile_user_and_group_slugs)
```
This method is overridden in CaseListFilter.
Moreover, For clarity i think using CaseListFilter for case export changes and LocationRestrictedMobileWorkerFilter for form export changes would be correct since they are used for rendering the filters as well. 
Both are subclassed from ExpandedMobileWorkerFilter. Would update at the places where ExpandedMobileWorkerFilter is being used instead of the above 2.

esoergel replied:
sounds good

mkangia commented on corehq/apps/export/views.py:
```diff
@@ -1703,16 +1707,53 @@ def _check_export_size(self, export_instances, filters):
+        return export_filter, export_specs
+
 
 class BulkDownloadNewFormExportView(DownloadNewFormExportView):
```
No changes to BulkDownloadNewFormExportView. The changes below are for DownloadNewCaseExportView. Diff misguided you :)

esoergel replied:
Ah, drat!  You're right.
Well in that case, it looks like these two classes are the only direct subclasses of `GenericDownloadNewExportMixin`, so the changes could go there instead (assuming they *are* actual duplications and not just similar looking)

mkangia commented on corehq/apps/locations/models.py:
```diff
@@ -611,6 +619,23 @@ def sql_location(self):
+    def location_and_descendants(cls, location_ids):
+        locations = []
+        for location_id in location_ids:
+            location = cls.by_location_id(location_id)
```
It would be interesting to see in which case it would not find a location. It would raise an exception at next step anyway but `DoesNotExist` would make it explicit.
Would use `get` instead of `by_location_id`

mkangia replied:
Got rid of the method only thanks to the consolidation you suggested.

mkangia committed:
Code Review improvements - 
Use better approach for finding locations and their descendants
Use only location_safe decorator for views [skip ci]

esoergel commented:
Sheel and I just offlined and decided that no groups should be included for restricted users (we may add that back in later, but for now it's a hard problem without an obvious answer).

esoergel commented on corehq/apps/locations/models.py:
```diff
@@ -611,6 +619,23 @@ def sql_location(self):
+    @classmethod
+    def users_at_locations_and_descendants(cls, location_ids):
+        location_ids = SQLLocation.location_and_descendants_ids(location_ids)
+        return UserES().location(location_ids).run().hits
```
> This method does not belong here. Since this returns ES hits have moved it to UserES. Is that the right place for this?

I think a standalone function in `dbaccessors` (or `esaccessors` if you prefer), but I don't feel strongly about it.

> I would still retain the method location_and_descendants_ids and move this logic there. I think it would come handy in future. And should leave it under SQLLocation and not dbaccessors?

Maybe `LocationManager` or a dbaccessor?  The way I think about this is:

* `SQLLocation` is the place for methods which operate on an instance of `SQLLocation`, stuff like `boston.get_descendants()`.
* `SQLLocation.objects` is the place for things which don't rely on a specific location, but return one or more locations.  For couch models, we typically make these `@classmethod`s on the `Document` class, but Django models use the model manager, so that's the convention I use for this stuff.  This is especially true for things like `accessible_to_user`, which is likely to be used as part of a more complex query.
* `dbaccessors` or `utils` are good places for things which feel more like single function calls than components of a querying library.

I'm looking at `dbaccessors.py` and `util.py` now, and I don't think we have a good distinction between the two.  I do see a bunch of "get users" functions involving locations in the locations dbaccessors file.

I just noticed `get_locations_and_children` in `util.py`, and it's identical to `location_and_descendants`.  I do think the model manager would also be a good place for that though, since it returns a queryset (and then you can easily slap on a `.location_ids()` at the end if that's what you want).

mkangia replied:
Thanks for all the experience learning Ethan. Really appreciate it.
Managers seem to be the correct place to go for data queries. Though that overlaps a bit with dbaccessors for me.
Yet to use more of utils but i see them more as methods given data to operate on them instead of fetching data.

Moving get_locations_and_children to LocationManager from util and using that instead of adding a new method.

mkangia commented on corehq/apps/export/forms.py:
```diff
@@ -184,42 +184,43 @@ def __init__(self, domain_object, *args, **kwargs):
-                ng_required='false',
-            ),
-            crispy.Div(
+        if not (hasattr(self, 'skip_layout') and self.skip_layout):
```
It sure is. Nice move.. thats cool!!!

Absolutely!! should add that as a class attribute with default False. Would also help add some documentation for its use.

mkangia commented on corehq/apps/reports/filters/api.py:
```diff
@@ -147,6 +148,16 @@ def get_groups(self, query, start, size):
         return [self.utils.reporting_group_tuple(g) for g in groups.run().hits]
 
 
+class LocationRestrictedEmwfOptionsView(EmwfOptionsView):
```
Would follow your instinct here since i am limited by knowledge here and would leave this as-is for now. And i think if some report uses this filter it would need to be location safe, comes from the name itself `LocationRestricted`. :)

mkangia committed:
Flatten list for group ids when exporting cases

mkangia committed:
Move common methods to GenericDownloadNewExportMixin

mkangia committed:
Move skip_layout as a class attribute with default False [skip ci]

dimagimon commented on corehq/apps/locations/models.py:
```diff
@@ -612,6 +620,14 @@ def sql_location(self):
         return self
 
 
+    @classmethod
```
E303 too many blank lines (2)

mkangia committed:
Dimagimon complaints

mkangia committed:
get_user_filter -> get_user_type_filter, use OR for form_filters [skip ci]

mkangia committed:
Reveal only accessible locations for case export, reveal no groups if user does not have Full Organization Access [skip ci]

mkangia committed:
Convert Query Set result to a list for location_and_descendant_ids

mkangia committed:
Show users assigned to web users location if not allowed Full Organisation access [skip ci]

dimagimon commented on corehq/apps/reports/filters/api.py:
```diff
@@ -10,12 +10,14 @@
 
 from corehq.apps.reports.filters.users import EmwfUtils
 from corehq.apps.es import UserES, GroupES, groups
+from corehq.apps.locations.dbaccessors import get_all_users_by_location
```
F401 'get_all_users_by_location' imported but unused

dimagimon commented on corehq/apps/reports/filters/case_list.py:
```diff
@@ -101,3 +117,8 @@ def get_sharing_groups(self, query, start, size):
+    def get_locations_query(self, query):
+        return (SQLLocation.active_objects
+                .filter_path_by_user_input(self.domain, query)
+                .accessible_to_user(self.request.domain, self.request.couch_user))
```
W292 no newline at end of file

esoergel commented on corehq/apps/reports/filters/api.py:
```diff
@@ -158,6 +158,18 @@ def get_locations_query(self, query):
+        if not self.request.can_access_all_locations:
+            user_location_id = self.request.couch_user.get_location_id(self.domain)
+            all_location_ids = SQLLocation.location_and_descendants_ids([user_location_id])
+            users = users.location(all_location_ids)
```
This can go in the base class (`EmwfOptionsView`).  If the request can't access all locations, then this should always be restricted.

mkangia replied:
Again there is a subclass to `EmwfOptionsView`, `LocationGroupFilter` which relies on `get_users` in the parent class. Thats why added these changes in the other two classes.

mkangia committed:
use list to convert ValuesListQuerySet to list [skip ci]

mkangia committed:
use decorator location_safe on classes instead of dispatch [skip ci]

mkangia committed:
move get_locations_and_children to LocationManager and add get_locations_and_children_ids using it to remove location_and_descendants_ids added to SQLLocation

mkangia committed:
Render static options for form and case export, case sharing groups for case exports only when user can access all locations

mkangia committed:
Default selection of assigned locations and their descendants for restricted user [skip ci]

mkangia committed:
1. Form Export : skip group and user type filter for restricted users, add scope filter for export for restricted user
2. Case Export : skip filtering by group ids, skip case sharing group ids for restricted users and populate only case sharing locations, add scope for export for restricted user [skip ci]

dimagimon commented:
There are 5 errors:

* corehq/apps/export/forms.py, line 485 - E126 continuation line over-indented for hanging indent
* corehq/apps/export/forms.py, line 487 - E126 continuation line over-indented for hanging indent
* corehq/apps/export/forms.py, line 490 - E126 continuation line over-indented for hanging indent
* corehq/apps/export/forms.py, line 492 - E126 continuation line over-indented for hanging indent
* corehq/apps/export/forms.py, line 648 - E127 continuation line over-indented for visual indent

mkangia committed:
Dimagimon complaints rested [skip ci]

mkangia committed:
Removed unwanted comment and remove group_ids from get_group_filter since they are not meant to be included and add comment to use GroupFormSubmittedByFilter instead [skip ci]

mkangia committed:
_get_group -> _get_group_ids

mkangia committed:
Move Users and locations filters along with group_ids fetch to a mixin and use that instead

mkangia committed:
fix tests failing due to call to full_delete on Location object and failing due to last_login on local machine

mkangia committed:
Reorder JS files for failing mocha test as a possible fix since the latter was imported in the former

mkangia committed:
extract users_at_locations_and_descendants query for testing and add test for query formed, Also use _get_es_user_types in EmwfFilterFormExport

mkangia committed:
Add tests for EmwfFilterExportMixin, EmwfFilterFormExport, FilterCaseESExportDownloadForm

mkangia committed:
include <script src="{% static 'reports/javascripts/reports.util.js' %}"></script> in mocha.html

dimagimon commented on corehq/apps/es/tests/test_esquery.py:
```diff
@@ -114,6 +114,40 @@ def test_filtered_forms(self):
+                                    {'terms': {'domain_memberships.assigned_location_ids': all_ids}}
+                                )
+                                }
+                            )}, {'term':{'is_active': True}},
```
E231 missing whitespace after ':'

dimagimon commented on corehq/apps/export/tests/test_export_forms.py:
```diff
@@ -41,3 +51,185 @@ def test_get_group_filter(self):
+        expected_filter = {
+            'terms':
+                {'form.meta.userID': self.user_ids
+                }
```
E124 closing bracket does not match visual indentation

dimagimon commented on corehq/apps/export/tests/test_export_forms.py:
```diff
@@ -41,3 +51,185 @@ def test_get_group_filter(self):
+        expected_filter = {
+            'terms':
+                {'form.meta.userID': self.user_ids
+                }
```
E124 closing bracket does not match visual indentation

mkangia committed:
dimagimon complaints rested [skip ci]

mkangia committed:
fix failing spec due to wrong ids

mkangia committed:
Code Refactor mainly around views added and LocationManagers

dimagimon commented on corehq/apps/es/users.py:
```diff
@@ -57,6 +57,30 @@ def show_only_inactive(self):
+
+    def user_ids_at_accessible_locations(self, domain, user):
+        all_users = self.users_at_accessible_locations(domain, user)
+        return [user['_id'] for user in all_users]
```
F812 list comprehension redefines 'user' from line 77

dimagimon commented on corehq/apps/locations/models.py:
```diff
@@ -332,12 +332,32 @@ def filter_path_by_user_input(self, domain, user_input):
         return self.get_queryset_descendants(direct_matches, include_self=True)
 
 
+    def get_locations(self, location_ids):
```
E303 too many blank lines (2)

mkangia committed:
Render only assigned locations and not descendants for default selection for restricted users

mkangia committed:
Add admin user type for form export filtering, Filter using user ids for user types in forms and case exports instead of using UserType and OwnerType Filter, Also take into account 'ProductData' and 'All Data' selection in case of case exports.

mkangia committed:
FilterFormESExportDownloadForm -> EmwfFilterExportMixin

mkangia committed:
LocationRestrictedEmwfOptionsView -> LocationRestrictedEmwfOptions, Move CaseListFilterOptions to api module, extract LocationRestrictedEmwfOptions Mixin [skip ci]

dimagimon commented on corehq/apps/export/tests/test_export_forms.py:
```diff
@@ -28,16 +37,194 @@ def test_get_datespan_filter(self):
+        self.assertEqual(self.filter_export._get_group_ids(group_ids_slug), group_ids)
+
+
+@patch('corehq.apps.reports.util.get_first_form_submission_received', lambda x: datetime.datetime(2015, 1, 1))
```
F811 redefinition of unused 'TestEmwfFilterFormExport' from line 21

dimagimon commented on corehq/apps/reports/filters/api.py:
```diff
@@ -10,12 +10,15 @@
 
 from corehq.apps.reports.filters.users import EmwfUtils
 from corehq.apps.es import UserES, GroupES, groups
+from corehq.apps.locations.models import LocationType, SQLLocation
```
F811 redefinition of unused 'SQLLocation' from line 11

mkangia committed:
reorder methods for better diff [skip ci]

mkangia committed:
Move _get_es_user_types to EmwfFilterExportMixin

mkangia committed:
Remove unused import

mkangia committed:
Fix dimagimon complaints

mkangia committed:
es_user_filter -> dynamic_filter_class, new_export_filters -> dynamic_filters

mkangia committed:
Use GroupFormSubmittedByFilter when filtering by group ids instead of fetching groups user ids

mkangia committed:
Lots of documentation, better naming and minor improvements

dimagimon commented on corehq/apps/export/esaccessors.py:
```diff
@@ -23,11 +24,11 @@ def get_case_export_base_query(domain, case_type):
+def get_groups_user_ids(group_ids):
     q = (GroupES()
-            .doc_id(group_id)
+            .doc_id(group_ids)
```
E127 continuation line over-indented for visual indent

dimagimon commented on corehq/apps/export/forms.py:
```diff
@@ -400,38 +411,157 @@ def get_multimedia_task_kwargs(self, export, download_id):
+            return GroupFormSubmittedByFilter(group_ids)
+
+
+    def get_special_owner_ids(self, mobile, admin, unknown, demo, commtrack):
```
E303 too many blank lines (2)

mkangia committed:
fix dimagimon complaint [skip ci]

mkangia committed:
remove print statements [skip ci]

esoergel commented on corehq/apps/reports/filters/api.py:
```diff
@@ -151,6 +153,86 @@ def get_groups(self, query, start, size):
         return [self.utils.reporting_group_tuple(g) for g in groups.run().hits]
 
 
+class LocationRestrictedEmwfOptionsMixin(object):
```
[not blocking] I believe all of this can go in the base `EmwfOptionsView`, since if `can_access_all_locations`, the behavior is unchanged, and if not, the modified behavior would be correct.  I'm going to be using this modified EMWF soon to make the Submit History Report location safe, so I can also just merge the classes then if you prefer.

mkangia replied:
`EmwfOptionsView` is inherited by other classes. I am not confident about this.

esoergel commented on corehq/apps/reports/filters/case_list.py:
```diff
@@ -60,44 +58,7 @@ def _selected_group_entries(self, mobile_user_and_group_slugs):
-                  .size(size)
-                  .sort("name.exact"))
-        return map(self.utils.sharing_group_tuple, groups.run().hits)
+            return self._get_assigned_locations_default()
```
:+1:

esoergel commented on corehq/apps/reports/models.py:
```diff
@@ -80,6 +81,17 @@ class HQUserType(object):
     included_defaults = (True, True, True, True, False)
 
     @classmethod
+    def user_type_mapping(cls):
```
Is this still in use?  I don't see it referenced elsewhere.

mkangia replied:
Nice Catch.. I added this for using UserTypeFilter but ended up relying on user_ids for each user type later on. Would remove this.

I am still slightly unclear about not using UserTypeFilter for user_types, exports used it but CaseList fetched ids for user types. We should definitely follow one thing for the entire application if possible?

Do we not save user_type for all forms and cases?

esoergel replied:
Unfortunately no, it looks like we just save user_type for forms, it's in the [mapping file](https://github.com/dimagi/commcare-hq/blob/master/corehq/pillows/mappings/xform_mapping.py#L19-L19)

esoergel commented on corehq/apps/users/models.py:
```diff
@@ -2124,6 +2124,18 @@ def get_sql_location(self, domain):
+            return SQLLocation.objects.get_locations(loc_ids)
+        else:
+            return []
+
```
haha, I actually just did the [same thing](https://github.com/dimagi/commcare-hq/commit/e85868196739e516e2b6d789b2acf54c24f90fd0) (I think you need to merge in master to get them).  Great minds...

mkangia replied:
Housa!!! 
Great minds! Absolutely :)

Can i keep this version? For the naming using `assigned` and It reusing already defined methods like `get_locations`?

esoergel replied:
Yeah, it sounds fine to convert mine over (it wouldn't make sense to have both versions of it) as long as CommCareUser and WebUser both have the same method names and arguments.  Apparently CommCareUsers sometimes make web requests, and I've seen some issues where we expect a web user, so we call a WebUser method, but that breaks if it's a CommCareUser.

esoergel reviewed (commented):
I left a bunch of feedback, feel free to do now, after this is deployed, or never, as seems appropriate.  It looks good overall, keep me posted on QA!

mkangia commented on corehq/apps/export/views.py:
```diff
@@ -1707,37 +1722,95 @@ def _check_export_size(self, export_instances, filters):
+            self.request.domain,
+            self.request.couch_user)
+        )
+        accessible_user_ids = UserES().user_ids_at_locations(accessible_location_ids)
```
There is a catch. Actually, I need accessible_location_ids later for OwnerFilter, hence could not fetch just the users from locations.

esoergel replied:
Oh right, I see that now.

mkangia commented on corehq/apps/export/forms.py:
```diff
@@ -400,38 +411,190 @@ def get_multimedia_task_kwargs(self, export, download_id):
+                demo=HQUserType.DEMO_USER in user_types,
+                commtrack=False,
+            )
+            return FormSubmittedByFilter(user_ids)
```
Absolutely.. Didn't notice that in my little 2 mobile worked test application. Thanks Mr. Hunt.

mkangia commented on corehq/apps/export/esaccessors.py:
```diff
@@ -26,11 +27,11 @@ def get_case_export_base_query(domain, case_type):
-    return q.run().hits[0]['users']
+         .doc_id(group_ids)
+         .fields("users"))
+    return flatten_list([hit["users"] for hit in q.run().hits])
```
👍

mkangia commented on corehq/apps/es/users.py:
```diff
@@ -57,6 +57,34 @@ def show_only_inactive(self):
+    def user_ids_at_accessible_locations(self, domain, user):
+        all_users = self.users_at_accessible_locations(domain, user)
+        return [_user['_id'] for _user in all_users]
+
```
Agreed! Should not fetch doc when need just the ids. Let me make this better. I would prefer retaining the method `users_at_location` but update it to return query instance and then chain `exclude_source` later when called. The concern is if such method is added later.. it might fetch records differently from the method fetching user_ids.

Oops! I did it again... Sorry for adding methods at not the best possible place again. Will move them to better locations.

mkangia replied:
got rid of `users_at_locations_and_descendants_query` and updated the methods under UserES to work as instances of UserES and justify there places under it. Moved the ids methods in the same module `es.users`. Not sure if any other place like `util` would suit them.

mkangia committed:
Realign methods added under UserES to return query instances to justify there places under UserES, use values_list instead of flatten_list,

dimagimon commented on corehq/apps/export/forms.py:
```diff
@@ -9,9 +9,13 @@
-    OwnerTypeFilter, ModifiedOnRangeFilter
+    OwnerTypeFilter, ModifiedOnRangeFilter, FormSubmittedByFilter, NOT
 )
+from corehq.apps.es.users import UserES, user_ids_at_locations_and_descendants
```
F401 'UserES' imported but unused

dimagimon commented on corehq/apps/export/views.py:
```diff
@@ -89,6 +99,7 @@
 from corehq.apps.users.models import Permissions
 from corehq.apps.users.permissions import FORM_EXPORT_PERMISSION, CASE_EXPORT_PERMISSION, \
     DEID_EXPORT_PERMISSION, has_permission_to_view_report
+from corehq.apps.es.users import UserES, user_ids_at_accessible_locations, user_ids_at_locations
```
F401 'UserES' imported but unused

dimagimon commented on corehq/apps/export/views.py:
```diff
@@ -1707,37 +1722,94 @@ def _check_export_size(self, export_instances, filters):
+        # Filter to be applied in AND with filters for export for restricted user
+        # Restricts to forms submitted by users at accessible locations
+        accessible_user_ids = (user_ids_at_accessible_locations(
+                self.request.domain, self.request.couch_user
```
E126 continuation line over-indented for hanging indent

mkangia committed:
use UserType(mobile) for filtering mobile users in form exports instead of fetching ids since that would be a huge performance hit for large domains, remove user_type_mapping since its not needed anymore

mkangia committed:
Fix issue with export filters in case of multiple filters by flatten_list, also update other places for same implementation

mkangia committed:
dimagimon complaints

mkangia committed:
fix errors introduced during optimisation and review fixes

esoergel commented:
Cool, looks good!  This is on staging now, right?

mkangia committed:
Fix failing already added tests

mkangia committed:
Add test cases for UserES#users_at_locations and #users_at_accessible_locations

mkangia committed:
extract _check_user_location_query for query-json test

mkangia committed:
Add tests for _get_es_user_types and _get_group_filter

dimagimon commented on corehq/apps/export/tests/test_export_forms.py:
```diff
@@ -2,42 +2,225 @@
+        }
+        self.assertEqual(group_filter.to_es_filter(), expected_filter)
+
+@patch('corehq.apps.reports.util.get_first_form_submission_received', lambda x: datetime.datetime(2015, 1, 1))
```
E302 expected 2 blank lines, found 1

mkangia committed:
Add tests for EmwfFilterFormExport _get_user_type_filter

mkangia committed:
Add test cases for get_case_filter for project and all data

mkangia committed:
add test cases for ExpandedMobileWorkerFilter, LocationRestrictedMobileWorkerFilter and CaseListFilter

dimagimon commented on corehq/apps/export/tests/test_export_forms.py:
```diff
@@ -2,42 +2,312 @@
+
+    @patch('corehq.apps.reports.filters.case_list.CaseListFilter.show_project_data', return_value=True)
+    @patch('corehq.apps.reports.filters.case_list.CaseListFilter.show_all_data', return_value=False)
+    @patch('corehq.apps.reports.filters.case_list.CaseListFilter.selected_user_types', return_value=[HQUserType.ADMIN])
```
E501 line too long (119 > 115 characters)

dimagimon commented on corehq/apps/export/tests/test_export_forms.py:
```diff
@@ -2,42 +2,312 @@
+
+    @patch.object(FilterCaseESExportDownloadForm, '_get_group_independent_filters', lambda x, y, z: [])
+    def test_get_filters_from_slugs_for_all_locations_access(self, case_sharing_locations_ids_patch,
+                                                             case_sharing_groups_patch, static_user_ids_for_group_patch,
```
E501 line too long (120 > 115 characters)

mkangia committed:
test_export_forms improvements : use form, filter as objects for patches on test classes, use subject for marking target class/object, use setUpClass for faster setup

mkangia committed:
Mark Form and Case list options as location safe for Data tab to be visible for restricted users, some minor improvements

dimagimon commented on corehq/apps/export/views.py:
```diff
@@ -11,13 +11,23 @@
 from django_prbac.utils import has_privilege
 from django.utils.decorators import method_decorator
-import json
+import json, re
```
E401 multiple imports on one line

mkangia committed:
move get_locations_and_children to LocationnManager for new code added

mkangia committed:
add get_assigned_sql_locations and get_assigned_sql_location_ids to CommCareUser and retain them in WebUser over get_location_ids and get_sql_locations

mkangia committed:
fix dimagimon complaint

mkangia commented:
@esoergel @sheelio :
I can finally say that this is done from my end now. Would wait for a couple of hours if you have any last-quick feedback and then move this to QA.

mkangia committed:
revert changes to calls for sql_locations and get_location_ids on CouchUser

mkangia committed:
retain versions of get_sql_locations and get_location_ids already present

mkangia committed:
Redo : add get_assigned_sql_locations and get_assigned_sql_location_ids to CommCareUser and retain them in WebUser over get_location_ids and get_sql_locations

mkangia committed:
Revert : Redo : add get_assigned_sql_locations and get_assigned_sql_location_ids to CommCareUser and retain them in WebUser over get_location_ids and get_sql_locations
And rename get_assigned_sql_locations and get_assigned_location_ids to existing method names since the build fails weirdly on those names

mkangia committed:
fix signature for get_sql_locations in WebUser and CommCareUser, remove duplicate definitions for get_location_ids and get_sql_locations in WebUser

mkangia committed:
revert Revert : Redo : add get_assigned_sql_locations and get_assigned_sql_location_ids to CommCareUser and retain them in WebUser over get_location_ids and get_sql_locations And rename get_assigned_sql_locations and get_assigned_location_ids to existing method names since the build fails weirdly on those names

mkangia committed:
Skip adding OR filter in case no filters provided for form exports

mkangia committed:
Update form and case export to reveal locations assigned to user and not just primary locations

mkangia committed:
Make create form and case export views location safe

esoergel commented on corehq/apps/locations/models.py:
```diff
@@ -274,6 +274,15 @@ def couch_locations(self, wrapped=True):
+        if not assigned_location_ids:
+            return self.none()  # No locations are assigned to this user
+        return self.all() & SQLLocation.active_objects.get_locations_and_children(assigned_location_ids)
+
```
I think this should go in `accessible_to_user` instead, since every usage of that should also account for multiple locations.  That method predates multiple locations per user, but part of the reason I abstracted that out was so we could change the meaning of "accessible to user" in the future and have everything using it automatically updated.

mkangia replied:
@esoergel will this conflict with other uses for `accessible_to_user`. I remember @proteusvacuum added this for his set of changes as well? If this does not then i can surely move this to `accessible_to_user`.

esoergel replied:
I don't think it'll conflict - Farid moved `accessible_to_user` out of the model manager and into the mixin, but I don't remember seeing him adding multiple locations support (although I may have just missed it).

mkangia replied:
@esoergel This is the `LocationQueriesMixin` only. 
@proteusvacuum can you help confirm if this change would be aligned with what you did?

mkangia committed:
update accessible_to_user to use assigned_locations instead of just the primary location instead of adding assigned_to_user

mkangia committed:
Fix issue with Multimedia Export for forms with user types from filters

dimagimon commented on corehq/apps/export/forms.py:
```diff
@@ -394,44 +410,215 @@ def _get_datespan_filter(self):
-    def get_multimedia_task_kwargs(self, export, download_id):
-        kwargs = super(FilterFormESExportDownloadForm, self).get_multimedia_task_kwargs(export, download_id)
+    def get_multimedia_task_kwargs(self, export, download_id, mobile_user_and_group_slugs):
+        kwargs = super(EmwfFilterFormExport, self).get_multimedia_task_kwargs(export, download_id, mobile_user_and_group_slugs)
```
E501 line too long (127 > 115 characters)

mkangia committed:
Move locations just before users in sources to have user types and groups above locations since user types have a fixed and small set of options, Also include extra options only when user can access all locations

mkangia committed:
include locations for Case export irrespective of case sharing for locations

mkangia committed:
dimagimon complaints

mkangia committed:
Always include locations in filter for Case and Form Export

mkangia committed:
Redirect User to Form/Case Exports on login

Redirects a user to Case Or Form Exports page if a user has permission to
view them and cant access all locations

This takes care of redirecting users to apt page since Form And Case exports are now location safe and users with access to Form and Case might not have access to locations listing which was default earlier

mkangia committed:
Merge master

mkangia committed:
revert incorrect submodule update during merge

mkangia committed:
Remove default scope for case sharing groups/locations

Using default scope was taken from CaseList feature but in Case export
default scope should not be needed.

mkangia committed:
Include user selected as user for case and not just owner

dimagimon commented on corehq/apps/export/forms.py:
```diff
@@ -9,9 +9,13 @@
+from corehq.apps.reports.filters.case_list import CaseListFilter
+from corehq.apps.reports.filters.users import LocationRestrictedMobileWorkerFilter
 from corehq.apps.groups.models import Group
+from corehq.apps.locations.models import SQLLocation
```
F401 'SQLLocation' imported but unused

mkangia committed:
Remove unused import [skip ci]

mkangia committed:
Filter cases by locations descendants as well in export

mkangia committed:
Fix test cases failing due to recent changes

Fix test cases failing due to removing default scope for case exports,
include locations descendants as wellfor locations selected and
adding user filter using LastModifiedBy

dimagimon commented on corehq/apps/export/tests/test_export_forms.py:
```diff
@@ -2,41 +2,341 @@
+from corehq.apps.domain.models import Domain
+from corehq.apps.reports.filters.case_list import CaseListFilter
+from corehq.apps.groups.models import Group
+from corehq.apps.locations.models import SQLLocation
```
F401 'SQLLocation' imported but unused

mkangia committed:
remove unused import [skip ci]

mkangia committed:
update prelogin reference

mkangia committed:
Update Restict Access by Location title

benrudolph commented:
also we usually link the PR in the relevant submodule here so the merger knows they need to merge that too

mkangia commented:
@esoergel More than 100 commits and even more conversations... You said to start with something simple in Org's Initiative. :)

benrudolph commented:
haha this has been quite the battle :crossed_swords:

mkangia committed:
Fix set up failing due to running in dec month

mkangia committed:
Revert "Fix set up failing due to running in dec month"

This reverts commit d1635289b443399d80d1b408dde90ffbf6ae5341.

mkangia commented:
@czue : Timing, so close.. thanks

mkangia committed:
Merge branch 'master' into location_export

mkangia committed:
fix tests failing due to renamed method

mkangia committed:
merge master

mkangia committed:
get_assigned_location_ids -> get_assigned_locations, get_assigned_sql_locations -> get_sql_locations

esoergel commented on corehq/apps/users/models.py:
```diff
@@ -2162,6 +2162,18 @@ def get_sql_location(self, domain):
+        if loc_ids:
+            return SQLLocation.objects.get_locations(loc_ids)
+        else:
+            return []
```
I recently saw some feedback about this very same thing - the consensus was that this method should return an empty queryset instead of an empty list.  See https://github.com/dimagi/commcare-hq/pull/14019#discussion_r89427760 and https://github.com/dimagi/commcare-hq/pull/14020 for context.

mkangia replied:
Oh yes! absolutely..

esoergel commented on corehq/privileges.py:
```diff
@@ -117,7 +117,7 @@ def get_name_from_privilege(cls, privilege):
             CUSTOM_BRANDING: _("Custom Branding"),
             ROLE_BASED_ACCESS: _("Advanced Role-Based Access"),
-            RESTRICT_ACCESS_BY_LOCATION: _("Organization-based user management restrictions"),
+            RESTRICT_ACCESS_BY_LOCATION: _("Organization-based data export and user management restrictions"),
```
:tada:

esoergel reviewed (approved):
I have some remaining feedback, but it can all be addressed in a follow-up PR.

czue commented:
> @esoergel  merged 107 commits into master from location_export 12 hours ago

wow! congrats @mkangia

mkangia commented:
@czue its not deployed yet... Waiting for any celebrations till i am sure that it is deployed and won't be reverted...

mkangia commented on corehq/apps/locations/models.py:
```diff
@@ -281,15 +281,13 @@ def couch_locations(self, wrapped=True):
-            return self.none()  # No locations are accessible to this user
-        return self.all() & get_locations_and_children(assigned_location_ids)
+            return self.none()  # No locations are assigned to this user
+        return self.all() & SQLLocation.active_objects.get_locations_and_children(assigned_location_ids)
```
@esoergel Changing this to object. Though i don't see why we won't follow the same principle as [here](https://github.com/dimagi/commcare-hq/pull/13362#discussion-diff-90930237R353) to call for the calling manager and make this `self.get_locations_and_children`

esoergel replied:
`accessible_to_user` is in `LocationQueriesMixin`, so it's available as a queryset method AND as a manager method.  `get_locations_and_children`, conversely, is only available as a manager method, so it's not guaranteed to be available (it's only defined in the subclass).

`get_locations_and_children` needs to be a manager method under it's current implementation.  We could make it a queryset method if we adopted the strategy of `self.all() & ...` like is used for `accessible_to_user`, but it makes more sense to me as a method which *returns* a queryset (ie a manager method) than as a method which *filters* a queryset (a queryset method).
